Enable support for decompression of compressed response within RestHighLevelClient

Hakky54 opened the pull request:
This pull request is to enable the RestHighLevelClient to handle an Elasticsearch response which has compressed content. It was already possible to do this with the low level client, see here [AsyncElasticsearchCompressedRequest](https://gist.github.com/Hakky54/e818a2d8d7d8153cd0ef89ca4143d805)
Compression can enabled within a node configuration with the following property: `http.compression: true` Compression can be triggered by a request from a client. Therefor you also need to provide additional information within the header of the request to Elasticsearch if a client really wants to enable it. That is possible with the following RequestOptions:
```
RequestOptions.Builder requestOptions = RequestOptions.DEFAULT.toBuilder()
    .addHeader("Accept-Encoding", "gzip")
```

With these two properties Elasticsearch will return a gzip compressed response body. The RestHighLevelClient can send a request with the above request options but it couldn't handle the response yet. This feature request contains the option to handle a compressed response if it is compressed with gzip.

Hakky54 committed:
Added decompression of gzip when gzip value is return as an header from Elasticsearch

Hakky54 committed:
Importing only specific classes

Hakky54 committed:
Merge pull request #1 from elastic/master

Pulling latest changes from master

Hakky54 committed:
Merge branch 'master' of github.com:Hakky54/elasticsearch into feature/add-support-for-decompression-of-compressed-response

Hakky54 committed:
Fixed imports

Hakky54 committed:
Fixed getting byte array the correct way

Hakky54 committed:
Fixed unit test

Hakky54 commented:
The other possible way to implement could be at this method level at line number 272 `org.elasticsearch.client.RestClient.convertResponse(InternalRequest request, Node node, HttpResponse httpResponse)` That would be less code duplicated compared to the already made changes within the pull request.

elasticmachine commented:
Pinging @elastic/es-core-features (:Core/Features/Java High Level REST Client)

Hakky54 committed:
Renamed variables and method names

andreidan reviewed (changes requested):
Thanks for using Elasticsearch @Hakky54 and for putting together this PR. Great catch on the lack of support for decompressing the response entity.

We use the Apache HTTP client that already has support for decompressing the http entity using the [GzipDecompressingEntity](http://hc.apache.org/httpcomponents-client-ga/httpclient/apidocs/org/apache/http/client/entity/GzipDecompressingEntity.html). I propose add support for decompressing the entity in the [`parseEntity` method](https://github.com/elastic/elasticsearch/blob/master/client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java#L1875) if the `Content-Encoding` header contains `gzip` using the apache 
`GzipDecompressingEntity` to wrap the `HttpEntity`.

I also propose we enhance the [`HttpCompressionIT`](https://github.com/elastic/elasticsearch/blob/master/qa/smoke-test-http/src/test/java/org/elasticsearch/http/HttpCompressionIT.java) with an integration test for the high level rest client. 

I believe the low level rest client should remain as it is for now, as that client puts the user in control with most things related to the HTTP interaction.

Thanks for working on this.

Hakky54 committed:
Applied review remarks

Hakky54 committed:
Merge branch 'master' into feature/add-support-for-decompression-of-compressed-response

Hakky54 commented:
Hi @andreidan thank you for your feedback and I am glad that you appreciate my pull request ! It looks like my initial solution was too verbose, I didn't know of the existing of the GzipDecompressingEntity. I have refactored it and pushed it here. I also enhanced the integration test within the HttpCompressionIT.  Please let me know what else could be improved :)

Hakky54 committed:
Reverted change

Hakky54 committed:
Removed unnecessary method as assertOk with StatusToXContentObject is sufficient

Hakky54 committed:
Removed unused import

Hakky54 committed:
Merge branch 'master' into feature/add-support-for-decompression-of-compressed-response

Hakky54 committed:
Removed duplicate assertions, as that part is the same as the unit test of testParseEntity()

andreidan commented on client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java:
```diff
@@ -1872,11 +1874,18 @@ protected final ElasticsearchStatusException parseResponseException(ResponseExce
+        final HttpEntity entity = Optional.ofNullable(httpEntity.getContentEncoding())
+            .map(Header::getValue)
+            .filter("gzip"::equalsIgnoreCase)
+            .map(headerValue -> (HttpEntity) new GzipDecompressingEntity(httpEntity))
```
super minor: shall we rename `headerValue` to `gzipHeaderValue` to point out we found the `gzip` encoding header?

Hakky54 replied:
Good suggestion, also applied!

andreidan reviewed (changes requested):
Thanks for iterating on this @Hakky54 This looks pretty good so far. 

I've left a few (mostly minor) suggestions. Thanks for working on this!

Hakky54 committed:
Applied review remarks

Hakky54 commented on client/rest-high-level/src/test/java/org/elasticsearch/client/RestHighLevelClientTests.java:
```diff
@@ -320,6 +326,50 @@ public void testParseEntity() throws IOException {
+    }
+
+    private static byte[] compress(String content) throws IOException {
+        return compress(content.getBytes(StandardCharsets.ISO_8859_1));
```
The reason why I used ISO_8859_1 here was because in a previous commit I was not using the GzipDecompressingEntity and somehow it was not able to correctly compress and decompress. I changed it now to UTF_8, great catch!

Hakky54 commented on qa/smoke-test-http/src/test/java/org/elasticsearch/http/HttpCompressionIT.java:
```diff
@@ -59,4 +66,29 @@ public void testUncompressedResponseByDefault() throws IOException {
+            .build();
+
+        SearchResponse searchResponse = restHighLevelClient().search(searchRequest, requestOptions);
+        assertOK(searchResponse);
```
Included asserting amount of hits and even the source of the hit

Hakky54 commented on client/rest-high-level/src/test/java/org/elasticsearch/client/RestHighLevelClientTests.java:
```diff
@@ -320,6 +326,50 @@ public void testParseEntity() throws IOException {
+        };
+
+        HttpEntity jsonEntity = new NByteArrayEntity(compress("{\"field\":\"value\"}"), ContentType.APPLICATION_JSON);
+        assertEquals("value", restHighLevelClient.parseEntity(enrichHeaderContentEncodingWithGzip(jsonEntity), entityParser));
```
Yes, you are right. It is indeed hard to read. I have refactored it with your suggestion, hope it make easier to understand and more enjoyable to read

Hakky54 commented on test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestCase.java:
```diff
@@ -139,6 +141,7 @@ public static boolean hasXPack() {
      * A client for the running Elasticsearch cluster
      */
     private static RestClient client;
+    private static RestHighLevelClient restHighLevelClient;
```
Good one, I refactored it!

Hakky54 committed:
Applied review remarks for excluding spied objects

Hakky54 commented on client/rest-high-level/src/test/java/org/elasticsearch/client/RestHighLevelClientTests.java:
```diff
@@ -320,6 +326,50 @@ public void testParseEntity() throws IOException {
+        return bos.toByteArray();
+    }
+
+    private static HttpEntity enrichHeaderContentEncodingWithGzip(HttpEntity httpEntity) {
```
Yes, also done. And instead of using BasicHttpEntity I have used NByteArrayEntity and included the content encoding and discarded enrichHeaderContentEncodingWithGzip method as that one is not needed anymore. Thank you for the feedback!

Hakky54 committed:
Merge branch 'master' into feature/add-support-for-decompression-of-compressed-response

andreidan commented:
@elasticmachine ok to test

andreidan commented:
@Hakky54 this looks great, thanks for iterating on it! I have one final suggestion and then we can merge this. Looking at the package structure a bit more I think we should put the HttpCompressionIT back and let it test just the low level rest client and have a HighLevelRestClientCompressionIT in the package where you moved HttpCompressionIT, which only checks the high level rest client (the `assertOk` from EsTestCase can then be local to the new HighLevelRestClientCompressionIT test).  

Sorry for the back and forth on this but I think it'd be an overengineer to create a separate test module in `elasticsearch/client/test`. 

Thanks for working on this, I think we'll be able to merge this in no time.

Hakky54 committed:
Applied review remarks

Hakky54 committed:
Merge branch 'master' into feature/add-support-for-decompression-of-compressed-response

Hakky54 commented:
Hehe, thank you!
It looks like you are not giving up on providing feedback, I would say keep on going! :D 

So what I did is:
- Rollback HttpCompressionIT to the initial state and placed it back to the original package
- Renamed HttpCompressionIT to HighLevelRestClientCompressionIT within the rest-high-level module
- dropped the additional assertOk method within the ESRestTestCase and inlined it within HighLevelRestClientCompressionIT

Hakky54 committed:
Added static import

andreidan reviewed (approved):
LGTM thanks for working on this @Hakky54

Hakky54 commented:
Thanks @andreidan It feels awesome to contribute back to the community after using Elasticseearch so long!

javanna commented:
Heya, I had left a comment on the original issue, see https://github.com/elastic/elasticsearch/issues/53555#issuecomment-601686565 . 

We use the async http client rather than the http client, although the former depends on the latter. We are free to use classes from the blocking variant, but they are not always a good fit. The discussion in https://issues.apache.org/jira/browse/HTTPCLIENT-1822 seems to suggest that more work is required to implement automatic decompression using the async primitives that the async client offers, and highlights the steps needed, that I suppose are valid no matter if implemented within the async http client or as a user of it.

I wonder if the implementation that we have introduced is equivalent, and why we have introduced it in our own client and not upstream for all the other users of async http client to benefit from.

andreidan commented:
@javanna thanks for raising the apache issue to our attention, I was not aware of it. I agree this would be nice to be handled by the underlying client (and the appropriate place for this) transparently but until there's support in the `httpclient` I think it's alright for us to support it.

We're handling the decompression when parsing the response, which will keep the networking async so I think this remains in line with the async client guarantees. Did we miss something in terms of being aligned with the async client invariants?

javanna commented:
> until there's support in the httpclient I think it's alright for us to support it.

Does this mean that we are going to contribute transparent content decompression back to apache http async client?

I guess it is ok to decompress this way after all for now. I would open the discussion on making automatic content decompression available in the low-level client: it may be possible to use the same approach given the limited set of features it exposes from the underlying async client, or we may have to first contribute back a solution to the async http client library. I don't think users expect to have to decompress stuff themselves. And backwards compatibility should not prevent us from making this improvement.

andreidan commented:
> Does this mean that we are going to contribute transparent content decompression back to apache http async client?

The way this is implemented at this stage, we can't port it back as it'll have to be a new implementation altogether as this implementation is ES HLRC specific. There isn't anyone in our team (afaik) actively pursuing this.

With regards to using automatic decompression in the LLRC I think you're correct in what the  options are.

Hakky54 commented:
Should I move the code changes to the LLRC or wait for now till this has been discussed within your team?

andreidan commented:
@Hakky54 sorry for the late reply. We discussed this today and agreed having this supported in the LLRC would be very useful, so if you'd like to continue working on this that would be very much appreciated.

Hakky54 commented:
Hi @andreidan that's awesome to hear! I couldn't push the changes to this pull request because I deleted the repository after the brach got merged. I applied the changes within a new feature branch and created a new pull request here: https://github.com/elastic/elasticsearch/pull/55413

Thanks @javanna for proposing to move this option to the llrc!

rand0m86 commented:
Hi @andreidan, I see you removed tag `backport pending` from this ticket so let me just double check - are there any plans to add this support to `6.8` branch?

You know I would be totally fine if the methods on `RestHighLevelClient` would not be `final` - it would be a way to override `parseEntity` method and add required functionality ourselves but what are the alternatives? Do you have a code snippet for achieving same result or we need to reimplement high level client?

rand0m86 commented:
For anyone else interested in getting GZIP working without update to version 7+ you need to tweak a few places.

The most important bit - response consumer, it should wrap `ContentBufferEntity` into `GzipDecompressingEntity`:
```java
/**
 * Basically a copy of {@link HeapBufferedAsyncResponseConsumer} with a support of gzip encoding.
 *
 * Once ES provides out of the box support for compression this class should be dropped.
 */
class GzipSupportingResponseConsumer extends AbstractAsyncResponseConsumer<HttpResponse> {
    private static final int DEFAULT_BUFFER_LENGTH = 4096;
    private final int bufferLimitBytes;
    private volatile HttpResponse response;
    private volatile SimpleInputBuffer buf;

    /**
     * Creates a new instance of this consumer with the provided buffer limit
     */
    GzipSupportingResponseConsumer(int bufferLimit) {
        if (bufferLimit <= 0) {
            throw new IllegalArgumentException("bufferLimit must be greater than 0");
        }
        this.bufferLimitBytes = bufferLimit;
    }

    /**
     * Get the limit of the buffer.
     */
    public int getBufferLimit() {
        return bufferLimitBytes;
    }

    @Override
    protected void onResponseReceived(HttpResponse response) {
        this.response = response;
    }

    @Override
    protected void onEntityEnclosed(HttpEntity entity, ContentType contentType) throws IOException {
        long len = entity.getContentLength();
        if (len > bufferLimitBytes) {
            throw new ContentTooLongException("entity content is too long [" + len +
                    "] for the configured buffer limit [" + bufferLimitBytes + "]");
        }
        if (len < 0) {
            len = DEFAULT_BUFFER_LENGTH;
        }
        this.buf = new SimpleInputBuffer((int) len, getByteBufferAllocator());
        ContentBufferEntity cbEntity = new ContentBufferEntity(entity, this.buf);
        this.response.setEntity(new GzipDecompressingEntity(cbEntity));
    }

    /**
     * Returns the instance of {@link ByteBufferAllocator} to use for content buffering.
     * Allows to plug in any {@link ByteBufferAllocator} implementation.
     */
    protected ByteBufferAllocator getByteBufferAllocator() {
        return HeapByteBufferAllocator.INSTANCE;
    }

    @Override
    protected void onContentReceived(ContentDecoder decoder, IOControl ioctrl) throws IOException {
        this.buf.consumeContent(decoder);
    }

    @Override
    protected HttpResponse buildResult(HttpContext context) {
        return response;
    }

    @Override
    protected void releaseResources() {
        response = null;
    }
}
```

Next, implement custom `HttpAsyncResponseConsumerFactory`:
```java
/**
 * Consumer factory which provides {@link GzipSupportingResponseConsumer} instead of
 * default {@link HeapBufferedResponseConsumerFactory} to enable GZIP support for ES calls.
 */
class CompressedHttpAsyncResponseConsumerFactory implements HttpAsyncResponseConsumerFactory {
    //default buffer limit is 100MB
    static final int DEFAULT_BUFFER_LIMIT = 100 * 1024 * 1024;

    @Override
    public HttpAsyncResponseConsumer<HttpResponse> createHttpAsyncResponseConsumer() {
        return new GzipSupportingResponseConsumer(DEFAULT_BUFFER_LIMIT);
    }
}
```

Next, use custom made `RequestOptions` for every high level client call:

```java
private RequestOptions createCustomRequestOptions() {
    RequestOptions.Builder builder = RequestOptions.DEFAULT.toBuilder();
    builder.setHttpAsyncResponseConsumerFactory(new CompressedHttpAsyncResponseConsumerFactory());
    return builder.build();
}

private void whatever(SearchRequest searchRequest) {
    SearchResponse response = client.search(searchRequest, createCustomRequestOptions()); // make sure you create options only once, there is no sense to recreate it on each call
}
```

And now you can enable GZIP on HTTP level:
```java
RestClientBuilder builder = RestClient.builder(httpHost);
builder.setHttpClientConfigCallback(clientBuilder -> {
    clientBuilder.addInterceptorFirst((HttpRequestInterceptor) (request, context) ->
        request.setHeader("Accept-Encoding", "gzip"));

    // Bonus time - compress all your request payloads, not only ES responses!
    clientBuilder.addInterceptorLast((HttpRequestInterceptor) (request, context) -> {
        if (request instanceof HttpEntityEnclosingRequest) {
            HttpEntityEnclosingRequest requestWithPayload = (HttpEntityEnclosingRequest) request;
            requestWithPayload.setEntity(new GzipCompressingEntity(requestWithPayload.getEntity()));
        }
    });
    return clientBuilder;
});
```

swallez commented:
@rand0m86 we have finally decided to backport this PR to the 6.8 branch. This has been done in PR #63087

rand0m86 commented:
Awesome news, thanks @swallez!

Once it's released I can really simplify my configuration code.
Do you have an idea when the next 6.8.x release version will be available?

Also, any plans to use my approach for compressing request payloads for #62044 (and backport it to 6.8) or do you need help in contributing to that one?

swallez commented:
@rand0m86 we are targeting a maintenance release for 6.8 in the near future, likely the end of October. So the wait shouldn't be very long.

About request payloads we're considering adding something like `RestClientBuilder.enableCompression()` that will enable compression for both requests and responses. Stay tuned!
